java/flight/flight-sql/src/main/java/org/apache/arrow/flight/sql/FlightSqlClient.java

```diff
@@ -0,0 +1,560 @@
+   * @param options RPC-layer hints for this call.
+   * @return a FlightInfo object representing the stream(s) to fetch.
+   */
+  public FlightInfo getTableTypes(final CallOption... options) {
+    final CommandGetTableTypes.Builder builder = CommandGetTableTypes.newBuilder();
+    final FlightDescriptor descriptor = FlightDescriptor.command(Any.pack(builder.build()).toByteArray());
+    return client.getInfo(descriptor, options);
+  }
+
+  /**
+   * Create a prepared statement on the server.
+   *
+   * @param query   The query to prepare.
+   * @param options RPC-layer hints for this call.
+   * @return The representation of the prepared statement which exists on the server.
+   */
+  public PreparedStatement prepare(final String query, final CallOption... options) {
+    return new PreparedStatement(client, query, options);
+  }
+
+  /**
+   * Helper class to encapsulate Flight SQL prepared statement logic.
+   */
+  public static class PreparedStatement implements AutoCloseable {
+    private final FlightClient client;
+    private final ActionCreatePreparedStatementResult preparedStatementResult;
+    private VectorSchemaRoot parameterBindingRoot;
+    private boolean isClosed;
+    private Schema resultSetSchema;
+    private Schema parameterSchema;
+
+    /**
+     * Constructor.
+     *
+     * @param client  The client. PreparedStatement does not maintain this resource.
+     * @param sql     The query.
+     * @param options RPC-layer hints for this call.
+     */
+    public PreparedStatement(final FlightClient client, final String sql, final CallOption... options) {
+      this.client = client;
+      final Action action = new Action(
+          FlightSqlUtils.FLIGHT_SQL_CREATEPREPAREDSTATEMENT.getType(),
+          Any.pack(ActionCreatePreparedStatementRequest
+                  .newBuilder()
+                  .setQuery(sql)
+                  .build())
+              .toByteArray());
+      final Iterator<Result> preparedStatementResults = client.doAction(action, options);
+
+      preparedStatementResult = FlightSqlUtils.unpackAndParseOrThrow(
+          preparedStatementResults.next().getBody(),
+          ActionCreatePreparedStatementResult.class);
+
+      isClosed = false;
+    }
+
+    /**
+     * Set the {@link #parameterBindingRoot} containing the parameter binding from a {@link PreparedStatement}
+     * operation.
+     *
+     * @param parameterBindingRoot a {@code VectorSchemaRoot} object containing the values to be used in the
+     *                             {@code PreparedStatement} setters.
+     */
+    public void setParameters(final VectorSchemaRoot parameterBindingRoot) {
+      if (this.parameterBindingRoot != null) {
+        if (this.parameterBindingRoot.equals(parameterBindingRoot)) {
+          return;
+        }
+        this.parameterBindingRoot.close();
+      }
+      this.parameterBindingRoot =
+          Objects.requireNonNull(parameterBindingRoot, "Parameter binding root cannot be null!");
```

Comment:
I guess there's an interesting question: why passing `null` should not invalidate the previous parameter?
But in that scenario, it probably the check can be done first (as it is done usually) and the following code should be equivalent:
```
  Objects.requireNonNull(parameterBindingRoot, "Parameter binding root cannot be null!");
  if (this.parameterBindingRoot.equals(parameterBindingRoot)) {
     return;
  }
  this.parameterBindingRoot.close();
  this.parameterBindingRoot = parameterBindingRoot;
```